File: backend/geomora_detect/scale_estimator.py
from __future__ import annotations

from typing import Any

from .models import DetectedElement

STANDARD_DOOR_HEIGHT_MM = 2100.0
STANDARD_SILL_HEIGHT_MM = 900.0
STANDARD_WINDOW_HEIGHT_MM = 1500.0
FACADE_MARGIN_FACTOR = 0.88


def estimate_scale(
    elements: list[DetectedElement],
    image_width: int,
    image_height: int,
    facade_bounds: list[int] | None = None,
) -> dict[str, Any] | None:
    if not elements or image_width <= 0 or image_height <= 0:
        return None

    doors = [element for element in elements if element.type == "door"]
    windows = [element for element in elements if element.type == "window"]

    wall_height_mm = None
    height_method = None
    height_confidence = 0.0

    if doors:
        door = max(doors, key=lambda element: element.confidence)
        door_height_norm = door.bbox_norm[3] - door.bbox_norm[1]
        if door_height_norm >= 0.08:
            wall_height_mm = STANDARD_DOOR_HEIGHT_MM / door_height_norm
            height_method = "door_height"
            height_confidence = min(0.9, door.confidence)

    if wall_height_mm is None and windows:
        window = max(windows, key=lambda element: element.confidence)
        sill_norm = max(1.0 - window.bbox_norm[3], 0.05)
        wall_height_mm = STANDARD_SILL_HEIGHT_MM / sill_norm
        height_method = "window_sill"
        height_confidence = min(0.75, window.confidence)

        window_height_norm = window.bbox_norm[3] - window.bbox_norm[1]
        if window_height_norm >= 0.08:
            alt_height = STANDARD_WINDOW_HEIGHT_MM / window_height_norm
            if alt_height > wall_height_mm * 0.75 and alt_height < wall_height_mm * 1.35:
                wall_height_mm = (wall_height_mm + alt_height) / 2.0
                height_method = "window_sill_height_blend"
                height_confidence = min(0.8, window.confidence)

    if wall_height_mm is None:
        return None

    wall_height_mm = _snap_mm(wall_height_mm, 50.0)
    wall_height_mm = max(2400.0, min(6000.0, wall_height_mm))

    x_min = min(element.bbox_norm[0] for element in elements)
    x_max = max(element.bbox_norm[2] for element in elements)
    span_norm = max(x_max - x_min, 0.25)
    span_norm = _extrapolate_span_norm(
        span_norm,
        window_count=len(windows),
        facade_bounds=facade_bounds,
        image_width=image_width,
    )

    aspect = image_width / image_height
    wall_length_mm = (span_norm * wall_height_mm * aspect) / FACADE_MARGIN_FACTOR
    wall_length_mm = _snap_mm(wall_length_mm, 100.0)
    wall_length_mm = max(4000.0, min(30000.0, wall_length_mm))

    return {
        "wall_length_mm": round(wall_length_mm),
        "wall_height_mm": round(wall_height_mm),
        "method": height_method or "openings_span",
        "confidence": round(height_confidence or 0.6, 3),
        "opening_span_norm": round(span_norm, 4),
    }


def _extrapolate_span_norm(
    span_norm: float,
    *,
    window_count: int,
    facade_bounds: list[int] | None,
    image_width: int,
) -> float:
    adjusted = span_norm

    if facade_bounds and len(facade_bounds) >= 4 and image_width > 0:
        facade_span = (facade_bounds[2] - facade_bounds[0]) / image_width
        adjusted = max(adjusted, facade_span * 0.85)

    if window_count <= 2:
        adjusted = max(adjusted, 0.72)
    elif window_count <= 4:
        adjusted = max(adjusted, 0.58)

    return min(adjusted, 0.98)


def _snap_mm(value: float, grid: float) -> float:
    return round(value / grid) * grid

File: plugin/geomora/core/scale_estimator.rb
# frozen_string_literal: true

module Geomora
  module Core
    class ScaleEstimator
      STANDARD_DOOR_HEIGHT_MM = 2100.0
      STANDARD_SILL_HEIGHT_MM = 900.0
      FACADE_MARGIN_FACTOR = 0.88

      def self.apply_hint!(params, scale_hint)
        return params unless scale_hint.is_a?(Hash)

        length = scale_hint['wall_length_mm']
        height = scale_hint['wall_height_mm']
        return params if length.to_f <= 0 || height.to_f <= 0

        params['wall_length'] = length.to_f.round
        params['wall_height'] = height.to_f.round
        params['scale_hint'] = scale_hint
        params
      end

      def self.from_detection(elements, image_width:, image_height:, facade_bounds: nil)
        return nil if elements.nil? || elements.empty?

        doors = elements.select { |element| element['type'] == 'door' }
        windows = elements.select { |element| element['type'] == 'window' }

        wall_height = estimate_wall_height(doors, windows)
        return nil unless wall_height

        wall_length = estimate_wall_length(
          elements,
          wall_height,
          image_width,
          image_height,
          window_count: windows.length,
          facade_bounds: facade_bounds
        )
        return nil unless wall_length

        {
          'wall_length_mm' => wall_length.round,
          'wall_height_mm' => wall_height.round,
          'method' => 'ruby_fallback',
          'confidence' => 0.5
        }
      end

      def self.estimate_wall_height(doors, windows)
        if doors.any?
          door = doors.max_by { |element| element['confidence'].to_f }
          bbox = door['bbox_norm']
          door_height_norm = bbox[3].to_f - bbox[1].to_f
          return snap_mm(STANDARD_DOOR_HEIGHT_MM / door_height_norm, 50.0) if door_height_norm >= 0.08
        end

        return nil if windows.empty?

        window = windows.max_by { |element| element['confidence'].to_f }
        bbox = window['bbox_norm']
        sill_norm = [1.0 - bbox[3].to_f, 0.05].max
        snap_mm(STANDARD_SILL_HEIGHT_MM / sill_norm, 50.0)
      end

      def self.estimate_wall_length(
        elements,
        wall_height,
        image_width,
        image_height,
        window_count: nil,
        facade_bounds: nil
      )
        return nil if image_width.to_i <= 0 || image_height.to_i <= 0

        x_min = elements.map { |element| element['bbox_norm'][0].to_f }.min
        x_max = elements.map { |element| element['bbox_norm'][2].to_f }.max
        span_norm = [x_max - x_min, 0.25].max
        span_norm = extrapolate_span_norm(
          span_norm,
          window_count: window_count || elements.count { |element| element['type'] == 'window' },
          facade_bounds: facade_bounds,
          image_width: image_width
        )

        aspect = image_width.to_f / image_height.to_f
        length = (span_norm * wall_height * aspect) / FACADE_MARGIN_FACTOR
        snap_mm(length.clamp(4000.0, 30_000.0), 100.0)
      end

      def self.extrapolate_span_norm(span_norm, window_count:, facade_bounds:, image_width:)
        adjusted = span_norm

        if facade_bounds.is_a?(Array) && facade_bounds.length >= 4 && image_width.to_i.positive?
          facade_span = (facade_bounds[2].to_f - facade_bounds[0].to_f) / image_width.to_f
          adjusted = [adjusted, facade_span * 0.85].max
        end

        if window_count <= 2
          adjusted = [adjusted, 0.72].max
        elsif window_count <= 4
          adjusted = [adjusted, 0.58].max
        end

        [adjusted, 0.98].min
      end

      def self.snap_mm(value, grid)
        ((value / grid).round * grid).round(1)
      end
    end
  end
end
